app/services/asset_preload_linker.rb

```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class AssetPreloadLinker
+  def self.append(response:, as:, url:, crossorigin: false, integrity: nil)
+    header = response.headers['Link'] || ''
+    header += ',' if header.present?
+    header += "<#{url}>;rel=preload;as=#{as}"
+    header += ';crossorigin' if crossorigin
+    header += ";integrity=#{integrity}" if integrity
```

Comment:
maybe a micro-optimization that isn't worth it, but `+=` will create and throw away all these intermediate strings? What if we appended as we went? And add `.dup` just incase the incoming `Link` string is frozen

```suggestion
    header = response.headers['Link']&.dup || String.new
    header << ',' if header.present?
    header << "<#{url}>;rel=preload;as=#{as}"
    header << ';crossorigin' if crossorigin
    header << ";integrity=#{integrity}" if integrity
```

Reply:
Micro-optimizations welcome! One of the advantages I was hoping for here is that there might be a small performance win, even just by skipping some of the options we have no need for (and those which are evaluated in the default implementation). One of the other things I was going to explore is if it's possible to switch the JavaScript rendering from `javascript_include_tag` to `tag.script` similar to the edits made in `stylesheet_helper.rb`.

This idea seems good. I might even go a step further and say we don't need to duplicate the header value and could write directly to it? Since we're assigning it anyways at the end of this method.

Reply:
> I might even go a step further and say we don't need to duplicate the header value and could write directly to it? Since we're assigning it anyways at the end of this method.

Something like...

```rb
def self.append(headers:, as:, url:, crossorigin: false, integrity: nil)
  if headers['Link']
    headers['Link'] << ','
  else
    headers['Link'] = ''
  end
  headers['Link'] << "<#{url}>;rel=preload;as=#{as}"
  headers['Link'] << ';crossorigin' if crossorigin
  headers['Link'] << ";integrity=#{integrity}" if integrity
end
```

Reply:
The `dup` is just in case something *else* sets a frozen string

And with `frozen_string_literal: true` at the top of the file, I think we need `String.new` instead of `''`

Reply:
Oh or we can use `+''` instead of `String.new` for a mutable string literal [(via StackOverflow)](https://stackoverflow.com/a/57795069)